app/upgrades/v4_1_0/upgrades_test.go

```diff
@@ -21,25 +24,70 @@ func TestUpgradeTestSuite(t *testing.T) {
+	vestingAddr := cVesting.GetAddress()
+	fmt.Printf("VestingAddr unvested: %+v\n", unvested)
+
+	accVestingBalance := s.App.BankKeeper.GetAllBalances(s.Ctx, vestingAddr)
+	fmt.Printf("Acc vesting bal: %s\n", accVestingBalance)
+
+	// create many validators to confirm the unbonding code works
+	newVal1 := s.SetupValidator(stakingtypes.Bonded)
+	newVal2 := s.SetupValidator(stakingtypes.Bonded)
+	newVal3 := s.SetupValidator(stakingtypes.Bonded)
+
+	// Delegate tokens of the vesting multisig account
+	s.StakingHelper.Delegate(vestingAddr, newVal1, sdk.NewInt(100))
+	s.StakingHelper.Delegate(vestingAddr, newVal2, sdk.NewInt(200))
+	s.StakingHelper.Delegate(vestingAddr, newVal3, sdk.NewInt(300))
 
-func dummyUpgrade(s *UpgradeTestSuite) {
-	s.Ctx = s.Ctx.WithBlockHeight(v4UpgradeHeight - 1)
-	plan := upgradetypes.Plan{Name: "v4.1.0", Height: v4UpgradeHeight}
-	err := s.App.UpgradeKeeper.ScheduleUpgrade(s.Ctx, plan)
+	// Undelegate part of the tokens from val2 (test instant unbonding on undelegation started before upgrade)
+	s.StakingHelper.Undelegate(vestingAddr, newVal3, sdk.NewInt(10), true)
+
+	// Redelegate part of the tokens from val2 -> val3 (test instant unbonding on redelegations started before upgrade)
+	_, err := s.App.StakingKeeper.BeginRedelegation(s.Ctx, vestingAddr, newVal2, newVal3, sdk.NewDec(1))
 	s.Require().NoError(err)
-	_, exists := s.App.UpgradeKeeper.GetUpgradePlan(s.Ctx)
-	s.Require().True(exists)
 
-	s.Ctx = s.Ctx.WithBlockHeight(v4UpgradeHeight)
+	// Confirm delegated to 3 validators
+	s.Require().Equal(3, len(s.App.StakingKeeper.GetAllDelegatorDelegations(s.Ctx, vestingAddr)))
+
+	// == UPGRADE ==
+	upgradeHeight := int64(5)
+	s.ConfirmUpgradeSucceeded(v4.UpgradeName, upgradeHeight)
+
+	// == VERIFICATION FEEBURN ==
+	feeBurnParam := s.App.FeeBurnKeeper.GetParams(s.Ctx)
+	s.Require().Equal("0", feeBurnParam.GetTxFeeBurnPercent())
+
+	// VERIFY MULTISIGN MIGRATION
+	accAfter := s.App.AccountKeeper.GetAccount(s.Ctx, vestingAddr)
+	_, ok := accAfter.(*vestingtypes.ContinuousVestingAccount)
+	s.Require().True(ok)
+
+	s.Require().Equal(1, len(s.App.BankKeeper.GetAllBalances(s.Ctx, vestingAddr)))
+	s.Require().Equal(4, len(s.App.StakingKeeper.GetAllDelegatorDelegations(s.Ctx, vestingAddr)))
+	s.Require().Equal(0, len(s.App.StakingKeeper.GetRedelegations(s.Ctx, vestingAddr, 65535)))
+
+	// check old multisign address balance
+	oldMultisigBalance := s.App.BankKeeper.GetAllBalances(s.Ctx, sdk.MustAccAddressFromBech32(v4.NotionalMultisigVestingAccount))
+	fmt.Printf("Old multisign address Upgrade Balance: %s\n", oldMultisigBalance)
+	totalDelegateBalance := s.App.StakingKeeper.GetDelegatorBonded(s.Ctx, sdk.MustAccAddressFromBech32(v4.NotionalMultisigVestingAccount))
+	fmt.Printf("old multisign address totalDelegateBalance %v\n", totalDelegateBalance)
+	s.Require().True(totalDelegateBalance.Add(oldMultisigBalance[0].Amount).GTE(unvested))
```

Comment:
This should be exact equal instead of GTE. 

unvested (total unvested amount) = totalDelegateBalance (from unvested amount) + oldMultisigBalance[0] (any unvested amount left) seems correct

local test works, so should be minor change

Reply:
updated